File: python/2016/d04.py
def load():
    with open('input') as f:
        return map(lambda x: x.strip(), f.readlines())

class Room:
    def __init__(self, s):
        tmp, checksum = s.split('[')
        parts = tmp.split('-')
        assert(checksum[-1] == ']')
        self.checksum = checksum[:-1]
        self.sector_id = int(parts[-1])
        self.name = '-'.join(parts[:-1])

    def __repr__(self):
        return '{}-{}[{}] {}'.format(self.name,
                                     self.sector_id,
                                     self.checksum,
                                     self.histogram())

    def histogram(self):
        unique = set(list(self.name))
        d = {letter: self.name.count(letter) for letter in unique}
        del d['-']
        return d

    def sorted_hist(self):
        def cmp((a, i), (b, j)):
            if i != j:
                return j - i
            return ord(a) - ord(b)
        return sorted(self.histogram().items(), cmp)

    def rc(self):
        return ''.join([c for (c,n) in self.sorted_hist()[:5]])

    def valid(self):
        return self.rc() == self.checksum

    def decrypt(self):
        def translate(c):
            if c == '-': return ' '
            x = ord(c) - ord('a')
            return chr((x + self.sector_id) % 26 + ord('a'))
        return ''.join([translate(c) for c in self.name])

def sector_if_valid(room):
    r = Room(room)
    if r.valid():
        return r.sector_id
    return 0

def part1(data):
    n = 0
    for line in data:
        n += sector_if_valid(line)
    return n

def part2(data):
    for line in data:
        r = Room(line)
        if r.valid():
            if r.decrypt().split()[0] == 'northpole':
                return r.sector_id
    return None

data = load()
print 'Part 1:', part1(data)
print 'Part 2:', part2(data)

File: src/Y2016/D04.hs
module Y2016.D04 where

import Data.Char (chr, ord)
import Data.List (intercalate, nub, sortBy)
import Data.List.Split (splitOn)
import Data.Maybe (fromMaybe)
import System.Environment (getArgs)
import Text.Read (readMaybe)

readSectorId :: String -> Int
readSectorId s = fromMaybe 0 (readMaybe s :: Maybe Int)

type Parsed = (String, Int, String)

parseLine :: String -> Parsed
parseLine room = (encrypted, sectorId, checksum)
    where parts = splitOn "[" room
          headParts = splitOn "-" $ head parts
          checksum = init $ last parts
          sectorId = readSectorId $ last headParts
          encrypted = intercalate "-" (init headParts)

rotate :: Int -> Char -> Char
rotate n c | c == '-' = ' '
           | otherwise = chr $ shifted + ord 'a'
     where normalized = ord c - ord 'a'
           shifted = (normalized + n) `mod` 26

decrypt :: Parsed -> String
decrypt (encrypted, sectorId, _) = map (rotate sectorId) encrypted

rc :: Parsed -> String
rc (encrypted, _, _) = take 5 $ map fst sorted
    where unique = nub $ filter (/= '-') encrypted
          counts = [length $ filter (== c) encrypted | c <- unique]
          sorted = sortBy cmp (zip unique counts)
          cmp (a, i) (b, j) | i /= j = compare j i
                            | otherwise = compare a b

sectorValue :: Parsed -> Int
sectorValue room@(_, sectorId, checksum)
    | rc room == checksum = sectorId
    | otherwise = 0

part1 :: [Parsed] -> Int
part1 ps = sum $ map sectorValue ps

part2 :: [Parsed] -> Int
part2 ps = sectorId
    where findNorthPole p = "northpole" == head (splitOn " " (decrypt p))
          (_, sectorId, _) = head $ filter findNorthPole ps

main :: IO ()
main = do
    [path] <- getArgs
    contents <- readFile path
    let parsed = map parseLine (lines contents)
    putStrLn $ "Part 1: " ++ show (part1 parsed)
    putStrLn $ "Part 2: " ++ show (part2 parsed)
